lls/cache.c

```diff
@@ -408,9 +474,15 @@ process_ping_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
 
 	num_granted_pkts = tb_ratelimit_allow_n(num_pkts, rs);
 
-	for (i = 0; i < num_granted_pkts; i++)
-		(*xmit_reply_fn)(iface, pkts[i]);
-	send_pkts(iface->id, tx_queue, num_granted_pkts, pkts);
+	while (num_ping_reply_pkts < num_granted_pkts && i < num_pkts) {
+		bool succ = (*xmit_reply_fn)(iface, pkts[i]);
+		if (succ)
+			ping_pkts[num_ping_reply_pkts++] = pkts[i];
+		else
+			rte_pktmbuf_free(pkts[i]);
+		i++;
+	}
+	send_pkts(iface->id, tx_queue, num_ping_reply_pkts, ping_pkts);
 
 	/* XXX #435: Adopt rte_pktmbuf_free_bulk() when DPDK is updated. */
 	for (i = num_granted_pkts; i < num_pkts; i++)
```

Comment:
Shouldn't we just keep the current value of `i` instead of updating it with `num_granted_pkts`? There's a problem right now: if `pkt[0]` is a malformed packet, `i = num_granted_pkts` will lead to a double deallocation since `i` is coming back.